--- a/BasicUsageEnvironment/BasicTaskScheduler.cpp
+++ b/BasicUsageEnvironment/BasicTaskScheduler.cpp
@@ -188,8 +188,12 @@
       if (mask == 0) mask = EVENT_TRIGGER_ID_HIGH_BIT;
 
 #ifndef NO_STD_LIB
-      if (fTriggersAwaitingHandling[i].test()) {
-	fTriggersAwaitingHandling[i].clear();
+      auto flag_was_set = [&](){
+        bool prev = fTriggersAwaitingHandling[i].test_and_set();
+        fTriggersAwaitingHandling[i].clear();
+        return prev;
+      };
+      if (flag_was_set()) {
 #else
       if (fTriggersAwaitingHandling[i]) {
 	fTriggersAwaitingHandling[i] = False;
